ch08performance/030misc-libs.ipynb.py

```diff
@@ -0,0 +1,385 @@
+#  - [Polars](https://docs.pola.rs): dataframes powered by a multithreaded, vectorized query engine, written in Rust
+#  - [boost-histogram](https://boost-histogram.readthedocs.io/en/latest/): Python bindings for the C++14 Boost::Histogram library
+#  - [Awkward Array](https://awkward-array.org/doc/main/): compiled and fast manipulation of JSON-like data with NumPy-like idioms
+#  - [Astropy](https://www.astropy.org): astronomy and astrophysics core library
+# 
+# JIT or AOT compiling libraries and the libraries providing Python bindings for a compiled
+# language are not mutually exclusive. For instance, PyTorch and Tensorflow offer users
+# "eager" and "graph" executation. Eager execution builds computational graph at runtime,
+# making it slow, easy to debug, and JIT compilable. On the other hand, "graph" execution
+# builds the computational graph at the kernel level, making it fast, hard to debug, and
+# with no need to JIT compile. Similarly, Awkward Array provides Python bindings for array
+# creation, but high-level operations on these arrays can be JIT compiled.
+# 
+# We will specifically look into Cython in the next lesson.
+# 
+# ## Distributed and parallel computing with Python
+# 
+# Along with the speed of execution of a program, scientific computation usually requires
+# a large amount of memory because of the massive nature of data produced at scientific
+# experiments. This issue can be addressed by employing distributed and parallel
+# computing frameworks to spread the computation across multiple machines or multiple
+# cores in a single machine. Parallel computing at a smaller level can be achieved by distributing
+# a task over multiple threads, but Python's GIL (Global Interpreter Lock) blocks
+# the interpretor from doing this for computational tasks. There are ongoing efforts, such
+# as the implementation of [PEP 703](https://peps.python.org/pep-0703/) to make GIL
+# optional, but Numba allows users to bypass GIL by pass `nogil=True` to `@jit`.
+# 
+# At a larger scale, Dask can be used to parallelize computation over multiple
+# machine or "nodes". Dask can even parallelize the code marked with `@numba.jit(nogil=True)`
+# to multiple threads in a machine, but it does not itself bypass the GIL.
+# 
+# ### Dask
+# 
+# Dask is a Python library for parallel and distributed computing. Dask supports
+# arrays, dataframes, Python objects, as well as tasks on computer clusters and
+# nodes. The API of Dask is very similar to that of NumPy, Pandas, and plain Python,
+# allowing users to quickly parallelize their implementations. Let's create a dummy
+# dataframe using dask.
+
+# %%
+import dask
+
+df = dask.datasets.make_people()  # returns a dask "bag"
+df = df.to_dataframe()
+
+# %%
+df
+# The computation gave us a dask object and not the actual answer. Why is that?
+#  We can visualise the dask task graph using -
```

Comment:
using ... is that a `-`?

Reply:
Ah I think I misplaced something here. Will fix it!